### PR TITLE
Improve reproduceability 2/3

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR allows one to pass CPU generators even when the model is run on GPU to improve the reproducability of models since this way the latents are not dependent on the CUDA / GPU hardware (see: https://github.com/pytorch/pytorch/issues/87992).

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_unclip.py [hunk 1]
```diff
@@ -293,7 +293,7 @@ def step(
                     " for the UnCLIPScheduler."
                 )
 
-            variance = variance * variance_noise
+            # variance = variance * variance_noise
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_unclip.py:296]: The only thing that seems to work here is to make the scheduler deterministic

File: src/diffusers/utils/torch_utils.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
 
 
-def torch_randn(
+def rand_tensorn(
```
Comment (patrickvonplaten) [inline src/diffusers/utils/torch_utils.py:29]: It seems like there is simply no way to make things fully deterministic even if we generate all tensors on CPU.

File: src/diffusers/models/unet_2d_condition.py [hunk 3]
```diff
@@ -387,6 +387,7 @@ def forward(
         # prepare attention_mask
         if attention_mask is not None:
             attention_mask = (1 - attention_mask.to(sample.dtype)) * -10000.0
+            attention_mask = attention_mask.to(sample.dtype)
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:390]:
```suggestion
```

File: src/diffusers/models/prior_transformer.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def __init__(
         self.proj_to_clip_embeddings = nn.Linear(inner_dim, embedding_dim)
 
         causal_attention_mask = torch.full(
-            [num_embeddings + additional_embeddings, num_embeddings + additional_embeddings], float("-inf")
+            [num_embeddings + additional_embeddings, num_embeddings + additional_embeddings], -10000.0
```
Comment (patrickvonplaten) [inline src/diffusers/models/prior_transformer.py:98]: For consistency, doesn't change the outputs here

File: tests/pipelines/unclip/test_unclip.py [hunk 5]
```diff
@@ -363,6 +362,41 @@ class DummyScheduler:
...
+        )
+
+        image = output.images[0]
+
+        np.save("/home/patrick_huggingface_co/diffusers-images/unclip/karlo_v1_alpha_horse_cpu.npy", image)
+        images = pipeline.numpy_to_pil(image)
+        images[0].save("/home/patrick_huggingface_co/diffusers-images/unclip/karlo_v1_alpha_horse_image.png")
+
+        assert image.shape == (256, 256, 3)
+        assert np.abs(expected_image - image).max() < 1e-1
```
Comment (patrickvonplaten) [inline tests/pipelines/unclip/test_unclip.py:397]: Even on CPU I cannot get a higher precision

Comment (patrickvonplaten) [reply tests/pipelines/unclip/test_unclip.py:397]: @anton-l I'd say we make this a nightly ? It takes 5 min on CPU

Comment (anton-l) [reply tests/pipelines/unclip/test_unclip.py:395]: Yeah, it should be ok to move the tests for less frequently updated pipelines completely to nightly IMO

File: tests/pipelines/unclip/test_unclip.py [hunk 6]
```diff
@@ -385,15 +417,19 @@ def test_unclip_karlo(self):
...
 
-        image = output.images[0]
+        image = np.asarray(pipeline.numpy_to_pil(output.images)[0], dtype=np.float32)
+        expected_image = np.asarray(pipeline.numpy_to_pil(expected_image)[0], dtype=np.float32)
+
+        # Karlo is extremely likely to strongly deviate depending on which hardware is used
+        # Here we just check that the image doesn't deviate more than 10 pixels from the reference image on average
+        avg_diff = np.abs(image - expected_image).mean()
 
+        assert avg_diff < 10, f"Error image deviates {avg_diff} pixels on average"
```
Comment (patrickvonplaten) [inline tests/pipelines/unclip/test_unclip.py:431]: Pretty high avg difference, but it doesn't seem like we can get closer than this.
